tokio/src/task/local.rs

```diff
@@ -288,6 +288,32 @@ struct LocalData {
     entered: Cell<bool>,
 }
 
+impl LocalData {
+    fn enter(&self, ctx: Rc<Context>) -> LocalDataEnterGuard<'_> {
+        let ctx = self.ctx.replace(Some(ctx));
+        let entered = self.entered.replace(true);
+
+        LocalDataEnterGuard {
+            local_data_ref: self,
+            ctx,
+            entered,
+        }
+    }
+}
+
+struct LocalDataEnterGuard<'a> {
```

Comment:
could we maybe add a comment on this explaining how it differs from the public `LocalEnterGuard`?  we now have two types in this module with similar names and similar behavior, one of which is private and the other of which is public. So, it would be nice to make the difference clear to readers. We may want to consider giving this type a different name, as well, if you can come up with something more descriptive.

Reply:
While writing the comment, I've realized that there is no need to separate `LocalEnterGuard` with `LocalDataEnterGuard`. So I've unified them!
Was it so radical?

Reply:
hmm, so one concern with combining the two guard types is that dropping a `LocalEnterGuard` will perform a second `CURRENT.with(...) call, rather than borrowing from the already-accessed thread-local as the `Reset` drop guard did. that means that if the `f` closure in `with` and `with_if_possible` panics and the thread-local gets torn down by unwinding *before* the `Drop` impl executes, we may attempt to access the thread local after its destruction, resulting in a double panic.

i'm unsure if this actually happens in practice or not, but it is a meaningful change in semantics relative to the previous implementation. if we're going to combine the two guard types, we should make sure that this case is tested and that we didn't introduce a double panic that didn't already exist.

Reply:
there may already be such a test, I'll have to double-check.

Reply:
Great point! But If I've understood the situation correctly, 
1. If a panic happens, local variables are dropped first, and then thread_local would be dropped later. for example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4dd7f70a8cdea6441775101582c1e6c3
2. Even though the variable CURRENT is being dropped, then aren't Cells inside it already being dropped? then double-panic problem should be reproduced in the previous code also.

Reply:
So I think it would not cause the problem, but if we have remaining worries, then it would be okay to separate again!

Reply:
I was considering about this, but combining two guards may not be related with the original problem.
So let me revert it. Thank you for pointing it! @hawkw

Reply:
Okay, I think the current implementation is fine. We can revisit this in a follow-up PR, potentially.